enterprise/server/remote_execution/runner/runner_test.go

```diff
@@ -643,6 +662,49 @@ func TestRunnerPool_TaskSize(t *testing.T) {
 	}
 }
 
+func TestRunnerPool_DelayedPause_PauseSkippedIfImmediatelyReused(t *testing.T) {
+	// Set a high pause delay to ensure runners are effectively never paused.
+	flags.Set(t, "executor.runner_pool.pause_delay", 100*time.Hour)
+
+	env := newTestEnv(t)
+	opts := *noLimitsCfg
+	opts.PoolOptions = &PoolOptions{ContainerProvider: FakeProvider}
+	pool := newRunnerPool(t, env, &opts)
+	ctx := withAuthenticatedUser(t, context.Background(), env, "US1")
+
+	r1 := mustGetNewRunner(t, ctx, pool, newTask())
+
+	mustAddWithoutEviction(t, ctx, pool, r1)
+
+	r2 := mustGetPausedRunner(t, ctx, pool, newTask())
+
+	assert.Same(t, r1, r2)
+	assert.Equal(t, 0, pool.PausedRunnerCount())
+
+	pauseCount := r1.Container.Delegate.(*fakeContainer).PauseCount
+	assert.Equal(t, 0, pauseCount, "should not have actually been paused due to delayed pause optimization")
+}
+
+func TestRunnerPool_DelayedPause_PausedInBackground(t *testing.T) {
+	// Set a short pause delay so that the runner is paused asynchronously but
+	// relatively quickly after being added to the pool.
+	flags.Set(t, "executor.runner_pool.pause_delay", 10*time.Millisecond)
+
+	env := newTestEnv(t)
+	opts := *noLimitsCfg
+	opts.PoolOptions = &PoolOptions{ContainerProvider: FakeProvider}
+	pool := newRunnerPool(t, env, &opts)
+	ctx := withAuthenticatedUser(t, context.Background(), env, "US1")
+
+	r1 := mustGetNewRunner(t, ctx, pool, newTask())
+	pauseCh := r1.Container.Delegate.(*fakeContainer).Paused
+
+	mustAddWithoutEviction(t, ctx, pool, r1)
+
+	// Should be paused after a short delay.
+	<-pauseCh
```

Comment:
Add `r2 := mustGetPausedRunner(t, ctx, pool, newTask())` to test that even if the runner is paused in the background, we can still successfully unpause it?

Reply:
good call - done